dnf json as a service

lavocatt opened the pull request:
This pull request includes:

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`
-->

teg commented on distribution/dnf-json.service:
```diff
@@ -0,0 +1,19 @@
+[Unit]
+Description=OSBuild dng-json service
+After=network.target
```
I don't think you need to order after the network target, but maybe I am missing something?

lavocatt replied:
That's copy and paste from `osbuild-remote-worker@.service` and `osbuild-worker@.service` I did not think about this too much. Should I remove this ?

teg replied:
I would remove it. This would make sure NetworkManager is started before the service (doesn't mean anything as the network might still be down) and stopped after it (which might be important to avoid network connection being dropped). However, we are not exposed to the network, so this should not be necessary. I would drop it. Unless someone corrects me ;) @gicmo @ondrejbudai ?

PS
I think it should not be in `osbuild-worker@.service` either, but in `-remote-` it makes sense.

teg commented on dnf-json:
```diff
@@ -1,21 +1,30 @@
+DNF_ERROR_EXIT_CODE = 10
 
-def timestamp_to_rfc3339(timestamp):
+def timestamp_to_rfc3339(timestamp, package):
```
This change looks right, but how did this work before?

teg replied:
If this is a bug fix, maybe making it a separate commit would be the correct thing to do.

lavocatt replied:
I the previous version the variable package was in the global scope thus accessible by the function. Since I moved the declaration to inside a class, the function was no longer able to access it, so I added it as a parameter.

teg commented on dnf-json:
```diff
@@ -123,98 +126,147 @@ def repo_checksums(base):
+    def _send(self):
+        self.client_address=('',)
+
+    def exit_with_dnf_error(self, kind: str, reason: str):
```
Should probably be `return_with_` rather than `exit_with_`, as we no longer exit the binary?

lavocatt replied:
It is solved by the next commit.

teg commented on dnf-json:
```diff
@@ -123,98 +126,147 @@ def repo_checksums(base):
+
+    def exit_with_dnf_error(self, kind: str, reason: str):
+        self._send()
+        self.send_response(500)
```
This preserves the behaviour (I think), so it is the right thing for the first version. But I would love to see this improved to return structured 400 errors in case of user error to explain all the different ways something might have gone wrong (conflicting requests, invalid repository, ...).

lavocatt replied:
Yes it does preserve the behaviour because we check in composer the value of the HTTP code. Before we were checking the value of the return code of DNF.

At the moment we already forward the reason why to the end user:

```
 time curl -k --key /etc/osbuild-composer/client-key.pem --cert /etc/osbuild-composer/client-crt.pem -H "Content-Type: application/json" -d @../tests/1477.json https://localhost/api/composer/v1/compose -i 
HTTP/1.1 400 Bad Request
Content-Type: application/json; charset=UTF-8
Date: Wed, 24 Nov 2021 17:02:46 GMT
Content-Length: 225

{"message":"Failed to depsolve requested package set: DNF error occured: MarkingErrors: Error occurred when marking packages for installation: Problems in request:\nmissing packages: jesuisunpaquetquinexistepas_idonotexist"}

real	0m0.754s
user	0m0.033s
sys	0m0.009s
```

teg replied:
I think this is right for now, but in the future we should make this a structured object so we can show the error message nicely in the UI. And we should distinguish between 400 (user made a mistake) and 500 (we made a mistake). Something like:
```
{
  "type": "MissingPackage",
  "payload": {
    "packages": [
      "jesuisunpaquetquinexistepas_idonotexist"
    ]
  }
}
```

Gundersanne replied:
Since depsolve is a job now  and asynchronous (as in, the request succeeds if the job was queued). So it wouldn't result in 400s/500s. But of course we still need a good way of distinguishing errors yea.

lavocatt replied:
Ok so I was not testing it with the v2. Now I'm doing it and I can see that indeed there's no way for a user to get feedback, at least through composer's api.

```
time curl -k --key /etc/osbuild-composer/client-key.pem --cert /etc/osbuild-composer/client-crt.pem -H "Content-Type: application/json" -d @../tests/compose_request_v2.json https://localhost/api/image-builder-composer/v2/compose -i 
HTTP/1.1 201 Created
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 Nov 2021 14:44:54 GMT
Content-Length: 113

{"href":"/api/image-builder-composer/v2/compose","kind":"ComposeId","id":"2989cd1f-5647-4783-ae58-6a2770cf89e1"}

real	0m0.043s
user	0m0.031s
sys	0m0.005s
┌─(✓)─[tlavocat@fedora]─[16% <─ ~/dev/tests]
└──╼> time curl -k --key /etc/osbuild-composer/client-key.pem --cert /etc/osbuild-composer/client-crt.pem -H "Content-Type: application/json" -d @../tests/compose_request_v2.json https://localhost/api/image-builder-composer/v2/composes/2989cd1f-5647-4783-ae58-6a2770cf89e1 -i 
HTTP/1.1 404 Not Found
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 Nov 2021 14:52:49 GMT
Content-Length: 201

{"href":"/api/image-builder-composer/v2/errors/21","id":"21","kind":"Error","code":"IMAGE-BUILDER-COMPOSER-21","operation_id":"21PoEE2eQFJMusCbAfSNVy9rXKv","reason":"Requested resource doesn't exist"}

real	0m0.041s
user	0m0.026s
sys	0m0.011s
```
How can we improve that ?

teg commented on dnf-json:
```diff
@@ -123,98 +126,147 @@ def repo_checksums(base):
+                        "version": package.version,
+                        "release": package.release,
+                        "arch": package.arch,
+                        "buildtime": timestamp_to_rfc3339(package.buildtime, package),
```
Why do we need to pass both `package` and `package.buildtime`?

teg commented on dnf-json:
```diff
@@ -123,98 +126,147 @@ def repo_checksums(base):
+class SystemDActivationSocketServer(socketserver.UnixStreamServer):
+    def server_bind(self):
+        log.debug("service bind")
+        if LISTEN_FDS == 0:
```
I think we may want to only support LISTEN_FDS=1 so we avoid confusion due to both this dameon and systemd "owning" the same socket.

lavocatt replied:
In the end we need to have this thing. Because sometime we want to be able to start dnf-json without systemd. This is the case when running in a container and for running the test cases generation.

teg commented on osbuild-composer.spec:
```diff
@@ -307,6 +307,7 @@ cd $PWD/_build/src/%{goipath}
 %doc README.md
 %{_mandir}/man7/%{name}.7*
 %{_unitdir}/osbuild-composer.service
+%{_unitdir}/dnf-json.socket
```
`osbuild-dnf-json.socket`

lavocatt commented on distribution/dnf-json.service:
```diff
@@ -0,0 +1,19 @@
+# systemd >= 240 sets this, but osbuild-worker runs on earlier versions
+Environment="CACHE_DIRECTORY=/var/cache/osbuild-dnf-json"
+StandardOutput=journal
+StandardError=journal
```
Indeed. I commited them by mistake ^^

teg commented on dnf-json:
```diff
@@ -159,98 +209,46 @@ class DnfJsonRequestHandler(BaseHTTPRequestHandler):
 
 log.info("Starting the dnf-json server")
 
 SOCKET_FILE_PATH = "/run/osbuild-dnf-json/api.sock"
```
Nitpick: I don't think this is used any more?

teg reviewed (approved):
Looks great to me! Maybe a review from @ondrejbudai and/or @Gundersanne would be great too, as this is quite delicate stuff.

Gundersanne commented on distribution/osbuild-dnf-json.service:
```diff
@@ -0,0 +1,17 @@
+Environment="CACHE_DIRECTORY=/var/cache/osbuild-dnf-json"
+
+[Install]
+WantedBy=osbuild-composer.service
```
Should be wanted by the workers too though, no @teg ? Or we include `Wants=` or `Require=` in workers and composer rather than WantedBy here? not sure

teg replied:
Good point! I didn't think of this. I don't think we need an install section at all, we should just require the socket from the service that depend on it. The service would be started on demand.

ondrejbudai replied:
Agreed with both of you, I suggest dropping the `Install` section and just directly using `Requires`, or `Wanted`. For now, I would just put `Requires` everywhere, it's imho the safest option.

lavocatt replied:
So should I make the worker require something ?

teg replied:
The worker should require the `.socket`, and that should be all that's needed, I think.

lavocatt replied:
Hmm, when I add the dependency to the `.socket` systemd refuses to start the process because it's missing a dependency

lavocatt replied:
And also, if the osbuil-dnf-json.service does not require its socket, the activation stops working.

ondrejbudai replied:
> Hmm, when I add the dependency to the .socket systemd refuses to start the process because it's missing a dependency

Can you post diff of your changes and the message from log?

> And also, if the osbuil-dnf-json.service does not require its socket, the activation stops working.

I think that the dnf-json.service always has to require its socket. That shouldn't be changed.

teg replied:
Both `osbuild-worker*.service` and `osbuild-dnf-json.service` should have `Requires=osbuild-dnf-json.socket`. I think that should be all that's needed.

lavocatt replied:
Do you think that `osbuild-remote-worker@.service` service too ? It is unclear because it does not have any requirements now.

teg replied:
Yeah, anything that will connect to the socket needs to `Require` it.

Gundersanne commented on internal/cloudapi/v1/v1.go:
```diff
@@ -175,9 +175,9 @@ func (h *apiHandlers) Compose(ctx echo.Context) error {
 		if depsolveResults.Error != "" {
 			if depsolveResults.ErrorType == worker.DepsolveErrorType {
-				return echo.NewHTTPError(http.StatusBadRequest, "Failed to depsolve requested package set: %s", depsolveResults.Error)
+				return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Failed to depsolve requested package set: %s", depsolveResults.Error))
```
Why add `fmt.Sprintf` ? Also cloudapi/v1 is deprecated (everything uses v2 now). It's also untested. I'd be ok with dropping any changes to cloudapi/v1 (in fact we sohuld just remove cloudapi/v1 soon, separate PR of course).

ondrejbudai replied:
Agreed.

lavocatt replied:
Oh yeah.. I really have to stop using v1 for my tests!

Gundersanne reviewed (commented):
Ah this looks so neat! I have a few comments/questions.

ondrejbudai commented on distribution/osbuild-dnf-json.socket:
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=OSbuild dnf-json socket
```
The tiniest nitpick ever: One unit has `OSBuild` in description, the other one has `OSbuild`.

ondrejbudai commented on internal/rpmmd/repository.go:
```diff
@@ -322,39 +324,31 @@ func runDNF(dnfJsonPath string, command string, arguments interface{}, result in
 	}
 
-	err = json.NewEncoder(stdin).Encode(call)
+	response, err := httpc.Post("http://unix", "application/json", strings.NewReader(string(bpost)))
```
`bytes.NewReader` would avoid a conversion

ondrejbudai reviewed (approved):
This is awesome! :-) `dnf-json` looks much cleaner after this change. I suggested several changes, up to you if you want to address them now, or in a follow-up.

Gundersanne commented on distribution/osbuild-worker@.service:
```diff
@@ -1,6 +1,6 @@
 [Unit]
 Description=OSBuild Composer Worker (%i)
-Requires=osbuild-local-worker.socket
+Requires=osbuild-local-worker.socket osbuild-dnf-json.socket
```
Also needs to be added in `osbuild-remote-worker@.service` and `osbuild-composer.service`; they all require dnf-json.

lavocatt commented:
Unfortunately I discovered that I have memory leaks to take care :/

thozza reviewed (commented):
This looks very interesting and I like where this is going! However I have a major issue with the current version.

Please correct me if I'm wrong, but I think that running `dnf-json` only as a service will break our tooling used to generate image test cases. Specifically `osbuild-pipeline` is not run as a service and it uses RPMMD with `dnf-json` to depsolve image package sets: https://github.com/osbuild/osbuild-composer/blob/8136e0f069efbae0d68cb1b6ad18913df7492ada/cmd/osbuild-pipeline/main.go#L121
https://github.com/osbuild/osbuild-composer/blob/8136e0f069efbae0d68cb1b6ad18913df7492ada/tools/test-case-generators/generate-test-cases#L66-L70
 
The issue is currently noticeable in the failing CI Manifest tests because they use the same approach as generators of image test cases.

Running `dnf-json` as a service probably makes sense, however personally I think that the generation of image test cases should be kept simple (executing a single command). This could mean that all tooling is modified to spin up `dnf-json` service in the background and stop it once it is not needed any more, or that `dnf-json` preserves the ability to be run as a command, not only as a service. I don't have a strong preference, but I think that this needs to be solved before the PR can be merged.

lavocatt commented on internal/rpmmd/repository.go:
```diff
@@ -322,39 +325,31 @@ func runDNF(dnfJsonPath string, command string, arguments interface{}, result in
 		arguments,
 	}
 
-	cmd := exec.Command(dnfJsonPath)
```
> This looks very interesting and I like where this is going! However I have a major issue with the current version.
> [...]
> Running `dnf-json` as a service probably makes sense, however personally I think that the generation of image test cases should be kept simple (executing a single command). This could mean that all tooling is modified to spin up `dnf-json` service in the background and stop it once it is not needed any more, or that `dnf-json` preserves the ability to be run as a command, not only as a service. I don't have a strong preference, but I think that this needs to be solved before the PR can be merged.

Thanks a lot! Yes you are right. I'm gonna change the tooling so that it uses the dameon version.

lavocatt commented:
> The issue is currently noticeable in the failing CI Manifest tests because they use the same approach as generators of image test cases.

The issue in the manifest test case is due to a memory leak in the daemon.

```
Nov 30 08:40:04 ip-10-30-18-53.us-east-1.aws.redhat.com kernel: Out of memory: Killed process 19731 (dnf-json) total-vm:3905420kB, anon-rss:3607436kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:7272kB oom_score_adj:0

```

lavocatt commented:
> Please correct me if I'm wrong, but I think that running `dnf-json` only as a service will break our tooling used to generate image test cases. Specifically `osbuild-pipeline` is not run as a service and it uses RPMMD with `dnf-json` to depsolve image package sets:

@thozza I've corrected the issue. I removed the dnfJsonPath from the api of RPMMD.
It works well also since I solved the memory leak issue (see https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1832645625). 

For clients of the Rpmmd api the behaviour is the same as before, because repository.go hides the complexity of calling the daemon through the unix socket.

SystemD is responsible for starting the process and I don't think that we should shut it down. Maybe we can add a timeout that closes the daemon if no communication was made to it for along time.

lavocatt commented:
My main issue now is that `create_repo_c` is creating a repository with access rights that does not allow the `dnf-json` daemon to access the files. I'll probably have to `chmod` stuff, but it seems hacky. I'll try to figure out something better.

ondrejbudai commented:
Isn't this a SELinux issue? IIRC SELinux prevents root processes to access home directories. A solution might be to run an ad-hoc HTTP server, I suggest something like `python -m http.server`, or a simple Go thing.

thozza commented:
> For clients of the Rpmmd api the behaviour is the same as before, because repository.go hides the complexity of calling the daemon through the unix socket.
> 
> SystemD is responsible for starting the process and I don't think that we should shut it down. Maybe we can add a timeout that closes the daemon if no communication was made to it for along time.

Right, this will work really nicely except for the case when generating image test cases...

The approach expects that RPMs which ship the `dnf-json` as a service are installed on the system when the API is used, e.g. when image test cases are being generated. This is in general not the case when generating image test cases, because thanks to `osbuild-pipeline` being run using `go run ...` we just need to generate manifests using the appropriate source code, but we don't need to have the RPMs built from it also installed on the system.

For CI testing, this should not be an issue, since the RPMs built from the tested source code are always installed on the system under test.

If you generate image test cases, e.g. using `./tools/test-case-generators/generate-all-test-cases ... remote ...`, the script by default installs latest versions of `osbuild` and `osbuild-composer` available in the distribution repositories. There is an option to define additional repos and these will have a higher priority. However there is a pretty unfortunate downside with `osbuild-composer`... we don't build RPMs for all architectures in our CI, for which we need to generate image test cases and test them (specifically `s390x` and `ppc64le`). This would be easily solved if we could use COPR repo which would contain the latest upstream builds (e.g. as it is done by PackIt for `osbuild` repository). However we don't have this yet for `osbuild-composer` and until the `osbuild-composer` RPMs with `dnf-json` as a service would get to Fedora 34/35 updates stable repo, generating image test cases for all architectures from the git `main` would be extremely harder.

Therefore if we go this way (having `dnf-json` only as a service), I propose that we first enable RPM builds for all architectures for Fedora using PackIt in osbuild-composer repository.

lavocatt commented:
> Isn't this a SELinux issue? IIRC SELinux prevents root processes to access home directories. A solution might be to run an ad-hoc HTTP server, I suggest something like `python -m http.server`, or a simple Go thing.

In fact, it's because `dnf-json` has a private `/tmp`, so when it tries to read the files at the path generated by the test it fails.
Using an `http.FileServer` in go makes the trick, good idea!

Gundersanne commented:
So dnf-json as a daemon does break composer as a standalone container for anything that expects the depsolving to happen in composer itself (so koji and the weldr api). This is the reason that the old-worker-new-composer regression test is broken as well, since there the composer PR version is running in a container and exposing the weldr api, but of course the dnf-json daemon isn't running in the container because we don't have systemd in the container.

We could expand `containers/osbuild-composer/entrypoint.py` to include dnf-json now, and invoke it the same way as we do weldr etc....

Gundersanne commented:
So I finally managed to get the regression test to work https://gitlab.com/osbuild/ci/osbuild-composer/-/jobs/1841343046, but with a hack https://github.com/osbuild/osbuild-composer/pull/2085/commits/3612c14c2579d169ca3da6107df5c19006511c86#diff-0ffbc497a008289d8efbb1d7edc7781d4e0c3a1e72ab964bd3a3a9401dadae86R257 not sure if there's a better way.

The fds are shared between the container entrpoint process and all the child processes (so osbuild-composer which has a variable number of fds, and dnf-json). So we somehow need to get the correct fd into dnf-json, which should of course default to 3.

thozza commented:
I had a look at https://github.com/osbuild/osbuild-composer/pull/2062/commits/8f39dd5c08291bf1973664647c7c593313fbad44 and think it won’t work with regenerating image test cases using `./tools/test-case-generators/generate-all-test-cases`. The reason is that the `dnf-json` installed on the system won’t be the one which creates a socket (until that version of composer gets to the distribution updates repos) and thus the RPMMD API from the “newer” sources copied over to the host won’t be able to connect to it.

I think that one solution could be to have a fallback in the RPMMD API to execute `dnf-json` as an executable if it is not listening on a socket, but I think this would be a bad approach as the code would be relevant for a limited time. So the alternative is IMO only to install the osbuild-composer version which includes `dnf-json` able to run as a service on the remote hosts when regenerating test cases. There are two possibilities:

1. Use PackIt and RPMs built in COPR when regenerating test cases by passing the COPR repo to the script. Passing the repo would be needed only for a limited time until the osbuild-composer version with `dnf-json` as a service would get to Fedora updates repo. 
2. Build the osbuild-composer RPMs on the remote host from copied sources and install them before regenerating test cases.

I'm in favor of (1), since having PackIt building composer RPMs for all arches would be definitely beneficial. (2) seems to me as adding additional potentially fragile step to generating test cases, but it could be a valid option. (2) would also require more coding as (1) is mostly about enabling PackIt and creating a YAML config for it.

ondrejbudai commented:
Note that COPR doesn't support s390x for el8 and both ppc64le and s390x for el9. For Fedora it supports all arches we need but that means we are stuck with Fedora for regenerating the test cases.

thozza commented:
> Note that COPR doesn't support s390x for el8 and both ppc64le and s390x for el9. For Fedora it supports all arches we need but that means we are stuck with Fedora for regenerating the test cases.

True.

We can ask COPR devels to enable these architectures, but s390x will be always only emulated 😞 .

But to be honest, I have been regenerating all image test cases on Fedora 34 for the past "forever" and it works great. There are sometimes issues which needed a workaround, but these show up in CI right away when a PR with regenerated test cases is proposed. So from my point of view, using only F34/F35 for regenerating test cases is not an issue...

thozza commented:
I implemented a workaround in the `generate-all-test-cases` script, which will allow the user to build the composer RPMs from sources on the runner and install them. Once merged, we should have a workaround for the transition period until composer RPMs with this change will get to the distribution repositories.

For more information see PR#2098

thozza commented on osbuild-composer.spec:
```diff
@@ -307,6 +307,7 @@ cd $PWD/_build/src/%{goipath}
 %doc README.md
 %{_mandir}/man7/%{name}.7*
 %{_unitdir}/osbuild-composer.service
+%{_unitdir}/osbuild-dnf-json.socket
```
This is a packaging bug. The `.socket` unit should be part of the `dnf-json` sub-package, as the `.service` file is.

lavocatt replied:
I don't understand how `.spec` is structured yet. I reproduced what was already done in the file for the sub-package `worker` and applied it to `dnf-json`. Why is the `.socket` unit not with the sub-package `worker`, is this a mistake too ? Should I correct the two ?

thozza replied:
No, those are fine. This is a bit misleading, but let me explain it.

In general the `.socket` unit starts a specific `.service` unit. The worker socket units are actually starting `osbuild-composer.service`, because it is the server's socket and the worker actually connects to it. So in this case this is correct. Worker does not listen on any socket, it connects to composer. In `dnf-json` case, the socket starts the service with the same base name and therefore it should be in the same binary RPM. In general everything would work just fine even now, because due to RPM dependencies the `dnf-json` sub-package will be pulled in when the main composer RPM is installed. But if you would install just the `dnf-json` sub-package, you would not be able to connect to it because systemd would not listen on the appropriate socket and would not start the service.

lavocatt replied:
Thanks a lot for the explanation!

thozza reviewed (commented):
In general, this looks very good. I found one nitpick and a packaging bug. I'm still testing the `generate-all-test-cases` with this PR...

lavocatt commented on cmd/osbuild-dnf-json-tests/main_test.go:
```diff
@@ -31,13 +36,13 @@ func TestFetchChecksum(t *testing.T) {
 	repoCfg := rpmmd.RepoConfig{
 		Name:      "repo",
-		BaseURL:   fmt.Sprintf("file://%s", dir),
+		BaseURL:   "http://localhost:9000",
```
The thing is, because the `dnf-json` service has its own private `/tmp` it could not access the files written at `dir` location. So I chose to keep the private `/tmp `and go this way for the test. I think the likelihood of a failure for the 9000 port is low, but I can enhance the log message in case the http server doesn't want to start to help a future debug.

thozza replied:
Thanks for explanation. this is fine!

thozza commented:
I tested generating image test cases with the proposed branch and `dnf-json` service does not get started (I didn't expect it to, since there is no code in `internal/rpmmd/repository.go` which would start it. The generation fails with:
```
panic: Could not depsolve: Post "http://unix": dial unix /run/osbuild-dnf-json/api.sock: connect: no such file or directory

goroutine 1 [running]:
main.main()
        /root/tmp.VhHjpMVJIn/sources/cmd/osbuild-pipeline/main.go:127 +0x11c6
exit status 2
```

I think that the solution could be to simply extend the `tools/test-case-generators/generate-test-cases` script to check if the `osbuild-dnf-json.socket` is active and if not, start it. This should not be an issue, since the script must be run with root privileges as it runs `osbuild` and `image-info`...

thozza commented on test/cases/generation.sh:
```diff
@@ -0,0 +1,42 @@
+sudo systemctl stop osbuild-dnf-json.socket
+
+# test the test cases generation
+WORKDIR=$(mktemp -d)
+mkdir -p "$WORKDIR"
```
This creates a temporary directory in `/tmp`, which is on `tmpfs` and based on the logs from the CI, there is not enough space for osbuild to store all of its data in the store directory. We usually use `/var/tmp`, which is on the disk and not in the RAM. I think that should solve the CI failure.

thozza commented on test/cases/generation.sh:
```diff
@@ -0,0 +1,42 @@
+mkdir -p "$WORKDIR"
+
+# test the test case generation when dnf-json socket is stopped
+sudo ./tools/test-case-generators/generate-test-cases\
```
You should add
```
OSBUILD_LABEL=$(matchpathcon -n $(which osbuild))
chcon $OSBUILD_LABEL tools/image-info
```
before calling the generator script.

For more information, check the "important note" in https://github.com/osbuild/osbuild-composer/tree/main/test#image-tests

thozza commented on tools/test-case-generators/generate-test-cases:
```diff
@@ -63,6 +67,16 @@ class TestCaseGenerator:
+                "systemctl is-active --quiet osbuild-dnf-json.socket")
+        if not is_subprocess_succeeding(command):
+            print("osbuild-dnf-json.socket is not running, starting it up")
+            command = shlex.split(
+                "systemctl start osbuild-dnf-json.socket --quiet")
```
The same comment as above about splitting the string.

thozza reviewed (commented):
Thanks for adding the test case for generating an image test case. 😍  I think that once the path for the osbuild store is changed in the test case, the CI could be green and the PR should be good to go...

thozza commented on test/cases/generation.sh:
```diff
@@ -0,0 +1,44 @@
+}
+
+# install requirements
+sudo dnf -q -y install make \
+    git \
+    go \
+    ranger \
+    openssl \
+    rpm-build \
+    aws \
+    qemu \
+    setools-console \
+    setroubleshoot \
+    container-selinux \
+    lvm2
```
We discussed this and I hoped that you'll eventually trim the list of packages only to those required for this test case to pass as there are definitely some unneeded packages. I'm not insisting that you do this in this PR, a follow up is fine by me. However if you are going to rebase this once more, you may as well clean it up.

I *think* that only `go` should be needed from the list. In addition, there is no reason to make dnf quiet by `-q` as we will at least see the installed package versions.

lavocatt replied:
Thanks @thozza Indeed we discussed it and I forgot to trim it. I'll do it in a follow up since I'm able to finally merge this PR. I'll keep you posted.

thozza reviewed (approved):
LGTM, thanks!
